### DISCUSSION TITLE
PorousFlow - Two phase injection

### DISCUSSION THREAD

**Discussion by wangshengwa** [idaholab/moose#22388] [Q&A Modules: General] [answered]
Hi all,
Following up with the problem described here #22251 .- where I have a two layered system (aquitard - aquifer) which is saturated with water. A borehole injects gas into the aquifer. 

I made a small MOOSE model to simulate this system. However, I have problems making the model converge. Looks like injection of water works fine, but convergence problems raise with the gas.

My code is here: https://github.com/wangshengwa/PorousFlow-stuff/tree/main/2pf_inj

Thanks I appreciate any help to make this work.
Cheers!
Wang

**Comment by GiudGiud** [comment]
Hello

Do you get convergence with LU?
I noticed you have a lu block set up.

Guillaume

**Reply by wangshengwa** [reply]
Thanks @GiudGiud , looks like I made a mistake in my dirackernels. Thanks!

**Comment by cpgr** [comment, marked as answer]
A couple of things that might help:

- I think that you need pgas >= pwater (at least I think that is the convention), so you probably want to initialise pgas = pwater
- You might need `fluid_phase = 1` in the borehole (phase 0 is water, phase 1 is gas)
- Likewise, `mass_fraction_component` should probably be 1 (the gas) in the borehole

Let me know if these help!

**Reply by wangshengwa** [reply]
Hi @cpgr 
Thanks a lot! I changed  `fluid_phase=1` and `mass_fraction_component=1`
I also reduced the tolerance to `nl_abs_tol = 1E-6`

Now it works!

However, I have numerical overshoots. I thought that KT  + superbee my help, but my model is still overshooting. Can you advise of how can I deal with this issue?

My model with KT stabilization: https://github.com/wangshengwa/PorousFlow-stuff/tree/main/2pf_kt

Cheers!
Wang

**Reply by cpgr** [reply]
Is the recorded wellbore flow oscillating, or is it something else? I haven't had a chance to look at it today, I'll try tomorrow.

**Reply by wangshengwa** [reply]
The pore pressure of the gas takes negative values of order 1E5. Thanks!

**Reply by cpgr** [reply]
I think it is probably an issue with your well, rather than something that could be fixed by stabilisation. @WilkAndy - any ideas?

**Reply by WilkAndy** [reply]
My suggestion is to copy the various similar examples in PorousFlow

https://mooseframework.inl.gov/modules/porous_flow/tests/dirackernels/dirackernels_tests.html
https://mooseframework.inl.gov/modules/porous_flow/sinks.html
https://mooseframework.inl.gov/modules/porous_flow/tests/fluidstate/fluidstate_tests.html
https://mooseframework.inl.gov/modules/porous_flow/tests/buckley_leverett/buckley_leverett_tests.html
https://mooseframework.inl.gov/modules/porous_flow/thm_example.html
